File: Kotlin/ch06/src/main/kotlin/ch06/BinarySearch.kt
package ch06

/**
 * 이진 탐색 (Binary Search)
 * Rust/ch06/src/search.rs → Kotlin 이디엄 포팅
 */
data class Point(val id: Int, val point: Double)

fun binarySearch(list: List<Point>, target: Double): Point? {
    var left = 0
    var right = list.size - 1
    while (left <= right) {
        val mid = (left + right) / 2
        when {
            target == list[mid].point -> return list[mid]
            target > list[mid].point -> left = mid + 1
            else -> right = mid - 1
        }
    }
    return null
}

fun main() {
    val data = listOf(
        Point(1, 94.73), Point(2, 121.44), Point(3, 176.23),
        Point(4, 224.72), Point(5, 671.78), Point(6, 877.88)
    ) // 이미 point 기준 정렬됨
    val found = binarySearch(data, 671.78)
    println("Found: id=${found?.id}, point=${found?.point}")
}

File: Rust/ch06/src/bin/binary_search.rs
// 이진 탐색 (Binary Search)
// Clang/06/BinarySearch 포팅.
use ch06::points_data::{data_set, Point};
use std::cmp::Ordering;

fn compare_point(a: &Point, b: &Point) -> Ordering {
    if a.point > b.point {
        Ordering::Greater
    } else if a.point < b.point {
        Ordering::Less
    } else {
        Ordering::Equal
    }
}

fn binary_search(list: &[Point], target: f64) -> Option<&Point> {
    let mut left: i64 = 0;
    let mut right: i64 = list.len() as i64 - 1;

    while left <= right {
        let mid = ((left + right) / 2) as usize;

        if target == list[mid].point {
            return Some(&list[mid]);
        } else if target > list[mid].point {
            left = mid as i64 + 1;
        } else {
            right = mid as i64 - 1;
        }
    }

    None
}

fn main() {
    let mut data = data_set();

    // 구매포인트에 대해 오름차순으로 정렬
    data.sort_by(compare_point);

    // 671.78 포인트 고객 찾기
    let found = binary_search(&data, 671.78).unwrap();

    println!("found... ID: {}, Point: {:.6} ", found.id, found.point);
}
